internal/rpc/digest_md5_integrity.go

```diff
@@ -0,0 +1,121 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"syscall"
+	"time"
+)
+
+type digestMD5IntegrityConn struct {
+	baseConnWrapper
+	buf bytes.Buffer
+
+	kic []byte
+	kis []byte
+
+	sendSeqNum int
+	readSeqNum int
+}
+
+func newDigestMD5IntegrityConn(conn net.Conn, kic, kis []byte) digestMD5Handler {
+	return &digestMD5IntegrityConn{
+		baseConnWrapper: baseConnWrapper{conn: conn},
+		kic:             kic,
+		kis:             kis,
+	}
+}
+
+func (d *digestMD5IntegrityConn) Write(b []byte) (n int, err error) {
+	if !d.writeDeadline.IsZero() && d.writeDeadline.Before(time.Now()) {
+		return 0, syscall.ETIMEDOUT
+	}
+
+	inputLen := len(b)
+	var seqBuf [4]byte
+	lenEncodeBytes(d.sendSeqNum, seqBuf[:])
+
+	outputLen := macDataLen + inputLen + macHMACLen + macMsgTypeLen + macSeqNumLen
+
+	n = 4
+	err = binary.Write(d.conn, binary.BigEndian, int32(outputLen-macDataLen))
+	if err != nil {
+		return
+	}
+
+	if n, err = d.conn.Write(b); err != nil {
+		return
+	}
+
+	hmac := getHMAC(d.kic, seqBuf[:], b)
+	if n, err = d.conn.Write(hmac); err != nil {
+		return
+	}
+
+	if n, err = d.conn.Write(macMsgType[:]); err != nil {
+		return
+	}
+
+	err = binary.Write(d.conn, binary.BigEndian, int32(d.sendSeqNum))
+	n += 4
+
+	d.sendSeqNum++
+	return
+}
+
+// Read will decode the underlying bytes and then copy them from our
+// buffer into the provided byte slice
+func (d *digestMD5IntegrityConn) Read(b []byte) (int, error) {
+	if !d.readDeadline.IsZero() && d.readDeadline.Before(time.Now()) {
+		return 0, syscall.ETIMEDOUT
+	}
+
+	n, err := d.buf.Read(b)
+	if len(b) == n || (err != nil && err != io.EOF) {
+		return n, err
+	}
+
+	var sz int32
+	binary.Read(d.conn, binary.BigEndian, &sz)
```

Comment:
You need to check the error here.

Reply:
This is also probably inefficient (it'll do a tiny unaligned read) but using a buffered read might be more trouble than it's worth (it'll muddle deadlines).

Reply:
I'm already basically doing a buffered read and muddling the deadlines, which is why we're tracking them separately on here, since you can't ask the `net.Conn` what it's current deadlines are